examples/java/src/main/java/org/apache/beam/examples/cookbook/DatastoreDelete.java

```diff
@@ -0,0 +1,101 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.examples.cookbook;
+
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.gcp.datastore.DatastoreIO;
+import org.apache.beam.sdk.io.gcp.datastore.DatastoreV1;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.options.Validation;
+import com.google.datastore.v1.Query;
+
+import javax.annotation.Nullable;
+
+/**
+ * An example that deletes entities from Cloud Datastore using DatastoreIO.
+ *
+ * <p>This example shows how to use DatastoreIO to delete entities from Cloud Datastore. Note that
+ * this example will write data to Cloud Datastore, which may incur charge for Cloud Datastore
+ * operations.
+ *
+ * <p>To run this example, users need to use gcloud to get credential for Cloud Datastore:
+ * <pre>{@code
+ * $ gcloud auth login
+ * }</pre>
+ *
+ * <p>To run this pipeline locally, the following options must be provided:
+ * <pre>{@code
+ *   --project=YOUR_PROJECT_ID
+ *   --kind=YOUR_DATASTORE_KIND_NAME
+ * }</pre>
+ *
+ * <p>To run this example using Dataflow service, you must additionally
+ * provide either {@literal --tempLocation} or {@literal --tempLocation}, and
+ * select one of the Dataflow pipeline runners, eg
+ * {@literal --runner=BlockingDataflowRunner}.
+ *
+ */
+public class DatastoreDelete {
+  /**
+   * Options supported by {@link DatastoreDelete}.
+   *
+   * <p>Inherits standard configuration options.
+   */
+  public static interface Options extends PipelineOptions {
+    @Description("Project ID for Cloud Datastore")
+    @Validation.Required
+    String getProject();
+    void setProject(String value);
+
+    @Description("Cloud Datastore Entity kind")
+    @Validation.Required
+    String getKind();
+    void setKind(String value);
+
+    @Description("Cloud Datastore Namespace")
+    String getNamespace();
+    void setNamespace(@Nullable String value);
+  }
+
+  public static void main(String[] args) {
+    Options options = PipelineOptionsFactory
+        .fromArgs(args).withValidation().as(DatastoreDelete.Options.class);
+
+    // A query to read all entities from the given kind.
+    Query.Builder query = Query.newBuilder();
+    query.addKindBuilder().setName(options.getKind());
+
+    // A transform to read entities
+    DatastoreV1.Read readEntities = DatastoreIO.v1().read()
+        .withProjectId(options.getProject())
+        .withQuery(query.build())
+        .withNamespace(options.getNamespace());
+
+    // A transform to delete entities.
```

Comment:
Summarizing what the Datastore team recommended for deletes,

- Read queries to a key-range (shard) should not be issued after the keys of the range were deleted, as it results in scanning the old keys in the key-range. However batched sequential cursor based read -> delete -> read -> delete... is good, as long as cursor is updated to not scan the old keys again.
- QuerySplitter returns queries that own one of more key-ranges but no two queries own the same key-range. This allows us to parallelize the deletes without worrying about the case above. 
- When worker dies after partially deleting a key-range, it will retry all the keys in the key-range, which is not desirable but still OK since worker failures are rare. 
- The other option is to break the fusion between Read and Write transform, but this results in copying all the data from datastore to dataflow, which is probably an overkill. 

Based on that, I think the example is good enough. @dhalperi let me know if you see any issue.